src/journal.rs

```diff
@@ -217,6 +239,50 @@ impl Journal {
         self.get_record()
     }
 
+    /// Wait for next record to arrive.
+    fn wait(&mut self, wait_time: JournalWait) -> Result<JournalWaitResult> {
+        let time = wait_time.map(usec_from_duration).unwrap_or(::std::u64::MAX);
+
+        match sd_try!(ffi::sd_journal_wait(self.j, time)) {
+            ffi::SD_JOURNAL_NOP => Ok(JournalWaitResult::Nop),
+            ffi::SD_JOURNAL_APPEND => Ok(JournalWaitResult::Append),
+            ffi::SD_JOURNAL_INVALIDATE => Ok(JournalWaitResult::Invalidate),
+            _ => Err(io::Error::new(InvalidData, "Failed to wait for changes"))
+        }
+    }
+
+    /// Wait for the next record to appear. Returns `Ok(None)` if there were no
+    /// new records in the given wait time.
+    pub fn await_next_record(&mut self, wait_time: JournalWait) -> Result<Option<JournalRecord>> {
+        match self.wait(wait_time)? {
+            JournalWaitResult::Nop => Ok(None),
+            JournalWaitResult::Append => self.next_record(),
+
+            // This is possibly wrong, but I can't generate a scenario with 
+            // ..::Invalidate and neither systemd's journalctl, 
+            // systemd-journal-upload, and other utilities handle that case.
+            JournalWaitResult::Invalidate => self.next_record()
```

Comment:
Ya, I'm not sure we can do any better in the context of this method. From the systemd manual:

> journal files were added or removed (possibly due to rotation). In the latter event, live-view UIs should probably refresh their entire display

Which isn't too clear about what changes from a `sd_journal` API perspective.